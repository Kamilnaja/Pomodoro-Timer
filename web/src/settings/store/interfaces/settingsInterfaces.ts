import { Settings } from '../../../../../types/settingsInterface';

export enum SortDirection {
  ASC = 'ASC',
  DESC = 'DESC',
}

export interface SettingsState {
  isLoading: boolean;
  error: any;
  settings: Settings;
  isCookieAcceptInfoVisible: boolean;
}
