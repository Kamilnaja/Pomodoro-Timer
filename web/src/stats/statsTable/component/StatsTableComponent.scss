.table {
  margin: 20px auto;
  max-width: 600px;
  table-layout: fixed;

  & &__head--day {
    display: flex;
    justify-content: space-around;
  }

  thead th {
    vertical-align: middle;
  }
}

.caption {
  caption-side: top;

  & &__header {
    display: inline-flex;
    font-size: 18px;
    margin-right: 20px;
  }
}
