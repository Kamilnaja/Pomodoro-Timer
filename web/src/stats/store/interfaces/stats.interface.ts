import StatsSearchResult from '../../../../../types/statistics.interfaces';

export default interface Stats {
  stats: StatsSearchResult;
  allPomodoros: number;
  error: string;
  isLoading: boolean;
}
