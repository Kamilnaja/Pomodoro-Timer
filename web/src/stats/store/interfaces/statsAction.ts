import { StatsActions } from "../actions/stats.actions";
import Stats from "./stats.interface";

export interface StatsAction {
  type: StatsActions;
  payload: Stats;
}
