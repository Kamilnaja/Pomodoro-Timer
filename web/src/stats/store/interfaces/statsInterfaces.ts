import { Result } from '../../../../../types/statisticsInterfaces';

export type HandleGetStats = (year: number, month: number) => void;

export type StatsComponentProps = { stats: Result[]; handleGetStats: HandleGetStats };

export type StatsContainerProps = {
  handleGetStats: HandleGetStats;
  stats: StatsState;
};

export interface StatsState {
  isLoading: boolean;
  error: string;
  results: Result[];
}
