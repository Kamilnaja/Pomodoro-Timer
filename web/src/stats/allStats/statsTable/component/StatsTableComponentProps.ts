import { PomodorosDoneInDay } from '../../../../../../types/statisticsInterfaces';
import { SortDirection } from '../../../../settings/store/interfaces/settingsInterfaces';

export interface StatsTableComponentProps {
  pageMonth: number;
  pageYear: number;
  pomodoros: PomodorosDoneInDay[];
  sortDirection: SortDirection;
  toggleSortDirection: () => void;
}
