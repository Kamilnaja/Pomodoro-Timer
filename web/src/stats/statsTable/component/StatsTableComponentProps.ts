import { Settings } from '../../../../../types/settingsInterface';
import { PomodorosDoneInDay } from '../../../../../types/statsInterfaces';
import { ToggleDisplayDirection, ToggleDisplayEmptyDays } from '../../../settings/store/actions/settingsActionsTypes';

export interface StatsTableComponentProps {
  pageMonth: number;
  pageYear: number;
  pomodoros: PomodorosDoneInDay[];
  settings: Settings;
  toggleDisplayDirection: ToggleDisplayDirection;
  toggleDisplayEmptyDays: ToggleDisplayEmptyDays;
}
