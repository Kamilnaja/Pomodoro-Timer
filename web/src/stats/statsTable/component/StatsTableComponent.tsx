import React from 'react';
import { Button, ButtonGroup, Table } from 'react-bootstrap';
import { PomodorosDoneInDay } from '../../../../../types/statisticsInterfaces';
import { StatsTableComponentProps } from './StatsTableComponentProps';
import './StatsTableComponent.scss';

export const StatsTableComponent = (props: StatsTableComponentProps) => {
  const daysInMonth = (month: number, year: number) => new Date(year, month, 0).getDate();

  const daysInCurrentMonth = () => daysInMonth(props.pageMonth + 1, props.pageYear);

  const findPomodorosInDay = (day: number): PomodorosDoneInDay =>
    props.pomodoros.find(v => new Date(v.date).getDate() === day);

  const getIndex = (i: number) => (props.sortDirection === 'DESC' ? daysInCurrentMonth() - i : i + 1);

  const longArr = (
    <>
      {Array.from(Array(daysInCurrentMonth()), (e, i) => (
        <tr key={i}>
          <td className="table__date">{getIndex(i)}</td>
          <td className="table__count">{findPomodorosInDay(getIndex(i))?.count}</td>
        </tr>
      ))}
    </>
  );

  const shortArr = (
    <>
      {props.pomodoros?.map((v, idx) => (
        <tr key={idx}>
          <td className="table__date">{v.date}</td>
          <td className="table__count">{v.count}</td>
        </tr>
      ))}
    </>
  );

  return (
    <Table className="stats__table table">
      <caption className="table__caption caption">
        <h2 className="caption__header">
          Results from: {props.pageMonth + 1}.{props.pageYear}
        </h2>

        <ButtonGroup size="sm">
          <Button onClick={() => props.toggleDisplayDirection()}>Toogle sort direction</Button>
          <Button>Show empty days</Button>
        </ButtonGroup>
      </caption>
      <thead className="table__head">
        <tr>
          <th>Day</th>
          <th>Number of pomodoros</th>
        </tr>
      </thead>
      <tbody>{longArr}</tbody>
    </Table>
  );
};
