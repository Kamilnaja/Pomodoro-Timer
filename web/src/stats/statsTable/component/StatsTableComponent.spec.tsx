import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { DisplayDirection } from '../../../settings/store/interfaces/settingsInterfaces';
import { createSettings } from '../../../settings/testing/settings.test.data';
import { createPomodorosDoneInDay } from '../../testing/stats.test.data';
import { daysInMonth, findPomodorosInDay, getPomodoroEntryAtIndex, StatsTableComponent } from './StatsTableComponent';
import { StatsTableComponentProps } from './StatsTableComponentProps';

describe('StatsTableComponent', () => {
  const pomodoros = createPomodorosDoneInDay();
  const pageMonth = 1;
  const pageYear = 2020;
  const toggleDisplayDirection = () => {};
  const wrapper = shallow(
    <StatsTableComponent
      pageYear={pageYear}
      pageMonth={pageMonth}
      toggleDisplayDirection={toggleDisplayDirection}
      toggleDisplayEmptyDays={() => {}}
      pomodoros={pomodoros}
      settings={createSettings()}
    />,
  );

  it('should match snapshot', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it.each([
    [0, 2020, 31],
    [1, 2020, 29],
    [2, 2020, 31],
    [6, 2018, 31],
    [1, 2021, 28],
  ])('daysInMonth should return correct number of days for month: %s, year: %s : %s', (month, year, numberOfDays) => {
    expect(daysInMonth({ pageMonth: month, pageYear: year } as any)).toBe(numberOfDays);
  });

  it.each([
    [
      10,
      {
        settings: { displayDirection: DisplayDirection.ASC },
        pageYear: 2020,
        pageMonth: 1,
      },
      11,
    ],
    [0, { settings: { displayDirection: DisplayDirection.ASC }, pageYear: 2020, pageMonth: 1 }, 1],
    [1, { settings: { displayDirection: DisplayDirection.ASC }, pageYear: 2020, pageMonth: 1 }, 2],
    [30, { settings: { displayDirection: DisplayDirection.ASC }, pageYear: 2020, pageMonth: 1 }, 31],
    [10, { settings: { displayDirection: DisplayDirection.ASC }, pageYear: 2020, pageMonth: 2 }, 11],
    [0, { settings: { displayDirection: DisplayDirection.ASC }, pageYear: 2020, pageMonth: 3 }, 1],
    [1, { settings: { displayDirection: DisplayDirection.ASC }, pageYear: 2020, pageMonth: 4 }, 2],
    [30, { settings: { displayDirection: DisplayDirection.ASC }, pageYear: 2020, pageMonth: 12 }, 31],
    [10, { settings: { displayDirection: DisplayDirection.DESC }, pageYear: 2020, pageMonth: 1 }, 19],
    [0, { settings: { displayDirection: DisplayDirection.DESC }, pageYear: 2020, pageMonth: 1 }, 29],
    [1, { settings: { displayDirection: DisplayDirection.DESC }, pageYear: 2020, pageMonth: 1 }, 28],
    [28, { settings: { displayDirection: DisplayDirection.DESC }, pageYear: 2020, pageMonth: 1 }, 1],
  ])('should return index for day: %s for props: %o', (index, props, res) => {
    expect(getPomodoroEntryAtIndex(index, props as StatsTableComponentProps)).toBe(res);
  });

  it.each([
    [
      10,
      {
        settings: {
          displayDirection: DisplayDirection.ASC,
        },
        pageYear: 2020,
        pageMonth: 1,
        pomodoros: [
          {
            date: '2020-01-10',
            count: 10,
          },
        ],
      },
      10,
    ],
    [
      0,
      {
        settings: {
          displayDirection: DisplayDirection.ASC,
        },
        pageYear: 2020,
        pageMonth: 1,
        pomodoros: [],
      },
      undefined,
    ],
    [
      21,
      {
        settings: {
          displayDirection: DisplayDirection.ASC,
        },
        pageYear: 2020,
        pageMonth: 1,
        pomodoros: [
          {
            date: '2020-01-20',
            count: 10,
          },
          {
            date: '2020-01-21',
            count: 3,
          },
        ],
      },
      3,
    ],
    [
      30,
      {
        settings: {
          displayDirection: DisplayDirection.ASC,
        },
        pageYear: 2020,
        pageMonth: 1,
        pomodoros: [
          {
            date: '2020-01-30',
            count: 30,
          },
        ],
      },
      30,
    ],
  ])('should return find number of pomodoros for index: %s, props: %o, result: %s', (index, props, count) => {
    expect(findPomodorosInDay(index, props as StatsTableComponentProps)?.count).toBe(count);
  });
});
