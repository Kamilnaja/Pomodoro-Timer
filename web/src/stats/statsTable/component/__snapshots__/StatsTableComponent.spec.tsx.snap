// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatsTableComponent should match snapshot 1`] = `
<ForwardRef
  bordered={true}
  className="stats__table table"
  hover={true}
  size="sm"
  striped={true}
>
  <caption
    className="table__caption caption"
  >
    <h2
      className="caption__header"
    >
      Results from: 
      2
      .
      2020
    </h2>
    <ButtonGroup
      role="group"
      size="sm"
      toggle={false}
      vertical={false}
    >
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Toogle sort direction
      </Button>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Show 
        all
         days
      </Button>
    </ButtonGroup>
  </caption>
  <thead
    className="table__head"
  >
    <tr>
      <th>
        Day
      </th>
      <th>
        Number of pomodoros
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="0"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        17
      </td>
    </tr>
    <tr
      key="1"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        15
      </td>
    </tr>
    <tr
      key="2"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        16
      </td>
    </tr>
    <tr
      key="3"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        12
      </td>
    </tr>
    <tr
      key="4"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        13
      </td>
    </tr>
    <tr
      key="5"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        16
      </td>
    </tr>
    <tr
      key="6"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        14
      </td>
    </tr>
    <tr
      key="7"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        18
      </td>
    </tr>
    <tr
      key="8"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        13
      </td>
    </tr>
    <tr
      key="9"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        13
      </td>
    </tr>
    <tr
      key="10"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        11
      </td>
    </tr>
    <tr
      key="11"
    >
      <td
        className="table__date"
      >
        2
      </td>
      <td
        className="table__count"
      >
        9
      </td>
    </tr>
    <tr
      key="12"
    >
      <td
        className="table__date"
      >
        NaN
      </td>
      <td
        className="table__count"
      >
        14
      </td>
    </tr>
    <tr
      key="13"
    >
      <td
        className="table__date"
      >
        NaN
      </td>
      <td
        className="table__count"
      >
        13
      </td>
    </tr>
    <tr
      key="14"
    >
      <td
        className="table__date"
      >
        NaN
      </td>
      <td
        className="table__count"
      >
        16
      </td>
    </tr>
    <tr
      key="15"
    >
      <td
        className="table__date"
      >
        NaN
      </td>
      <td
        className="table__count"
      >
        11
      </td>
    </tr>
    <tr
      key="16"
    >
      <td
        className="table__date"
      >
        NaN
      </td>
      <td
        className="table__count"
      >
        18
      </td>
    </tr>
    <tr
      key="17"
    >
      <td
        className="table__date"
      >
        NaN
      </td>
      <td
        className="table__count"
      >
        19
      </td>
    </tr>
    <tr
      key="18"
    >
      <td
        className="table__date"
      >
        NaN
      </td>
      <td
        className="table__count"
      >
        16
      </td>
    </tr>
    <tr
      key="19"
    >
      <td
        className="table__date"
      >
        NaN
      </td>
      <td
        className="table__count"
      >
        15
      </td>
    </tr>
    <tr
      key="20"
    >
      <td
        className="table__date"
      >
        NaN
      </td>
      <td
        className="table__count"
      >
        8
      </td>
    </tr>
  </tbody>
</ForwardRef>
`;
